docs/Migration Guide.md

```diff
@@ -0,0 +1,152 @@
+- A few other APIs have been renamed for clarity, such as
+  `JS_BufferIsCompilableUnit()` to `JS_UTF8BufferIsCompilableUnit()`.
+- If using a source hook to manage in-memory sources,
+  `js::SourceHook::load()` now has two out parameters for either UTF8
+  source code or wide char source code.
+  Use only one of them.
+  (None of these examples currently use source hooks.)
+- Your code will need to include `<js/SourceText.h>` for
+  `JS::SourceText`.
+
+### String encoding ###
+
+`JS_EncodeString()` has been replaced with `JS_EncodeStringToASCII()`
+and `JS_EncodeStringToUTF8()`, so that the encoding of the returned
+string is always clear.
+
+**Recommendations:**
+- Depending on what you were using `JS_EncodeString()` for, replace it
+  with either the ASCII or UTF8 version.
+  For better performance in cases where you know the string only
+  contains ASCII characters, use the ASCII version, otherwise use the
+  UTF8 version.
+- These functions now return a `JS::UniqueChars` smart pointer, so there
+  is no longer any need to `JS_free()` the return value.
+- Where possible, change any of your internal APIs where the return
+  value of `JS_EncodeString()` was returned, to also return a smart
+  pointer.
+
+### Hash tables ###
+
+It is no longer necessary to initialize `JS::GCHashMap<>` with `init()`.
+
+**Recommendation:** Simply remove the call to `init()`.
+
+### Principals ###
+
+Functions having to do with stack frames, such as
+`JS::GetSavedFrameSource()` and others, now take a `JSPrincipals*`
+pointer as an argument.
+**FIXME**: What is that for?
+
+**Recommendations:**
+- **FIXME**
+
+### Headers ###
+
+There are now more optional headers which should be included separately
+in code which uses their functionality.
+This is a list of common ones that might be used by embeddings, but in
+general if you are missing function definitions when compiling your
+code, try checking if you might have to include another header.
+
+- `JS::Compile()`, `JS::Evaluate()`, and similar functions —
+  `<js/CompilationAndEvaluation.h>`
+- `JS::ContextOptions` — `<js/ContextOptions.h>`
+- `JS::WarnUTF8()` and similar functions, and `JS::SetWarningReporter()`
+  — `<js/Warnings.h>`
+
+### Various API changes ###
+
+This is a non-exhaustive list of minor API changes and renames.
+
+- `JS_DefineProperty()` for a property with getter and setter function
+  objects no longer needs to use `JS_DATA_TO_FUNC_PTR()`.
+  There is an appropriate overload to use instead.
+- `JS_GetGlobalForObject()` → `JS::GetNonCCWObjectGlobal()`
+- `JS_NewArrayBuffer()` → `JS::NewArrayBuffer()`; ditto for
+  `JS_NewArrayBufferWithContents()` and `JS_NewExternalArrayBuffer()`.
+  In addition, `JS::NewExternalArrayBuffer()` has dropped support for
+  the ref callback.
+- `JS_ObjectIsFunction()` → `js::IsFunctionObject()`
+- `JS_ReportWarningASCII()` → `JS::WarnASCII()`
+- `JS_ReportWarningUTF8()` → `JS::WarnUTF8()`
+- `JS_updateMallocCounter()` → ??? **FIXME**
```

Comment:
Looks like it's JS::AddAssociatedMemory now.

Reply:
I couldn't really figure out how to use that without having it crash; in any case it seems not to be a drop-in replacement. (And could we add `JS::MemoryUse::Embedding`?)